### PR TITLE
Fix distributed autograd gradients synchronization

### PR DESCRIPTION
**Pull request by mrshenli**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57918 Fix distributed autograd gradient accumulation
* #57917 Fix Raw RPC/RRef Micro-batch-based Pipelining
* #57926 Use RPC context streams to cover serde ops
* #57896 Revert "Make RRefContext get devices from RPC agent when creating OwnerRRef (#57443)"
* **#57792 Fix distributed autograd gradients synchronization**

There are two problems when using CUDA RPC with distributed autograd
and distributed optimizer:

1) In local autograd engine, all autograd functions/nodes, including
AccumualteGrad will use the forward stream for backward computation.
But distributed autograd skips AccumulateGrad autograd function/node
and directly calls into `AccumulateGrad::accumulateGrad`. As the
result, it will use the default stream to accumulate gradients
instead of the forward stream. This commit changes that and uses the
forward stream to accumulate gradients, matching forward behavior.
2) Distributed optimizer and distributed autograd backward are
separate RPC calls, and CUDA streams are not synchronized across
different RPC calls. As a result, distributed optimizer might
consume gradients before they are ready. This commit uses CUDA
events to record the completion of gradient computation, and use
those events to block current streams when getGradients() are called.

Differential Revision: [D28274876](https://our.internmc.facebook.com/intern/diff/D28274876)

### CODE REVIEW

**Comment by mrshenli** [comment]
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28274876).

**Comment by mrshenli** [comment]
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28274876).

File: torch/csrc/distributed/autograd/context/container.cpp [hunk 1]
```diff
@@ -246,20 +246,19 @@ void DistAutogradContainer::sendReleaseContextRpc(
           options);
 
       std::weak_ptr<rpc::JitFuture> wp = cleanupFuture;
-      cleanupFuture->addCallback(
```
Comment (mrshenli) [inline torch/csrc/distributed/autograd/context/container.cpp:249]: this is clang format change only

**Comment by mrshenli** [comment]
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28274876).

**Comment by mrshenli** [comment]
Is there another problem with CUDA RPC + dist autograd when micro-batch is enabled. Micro-batch basically will launch multiple RPC calls on the param in the forward pass, which is fine. However, in the backward pass, it means that the same gradient will also be updated multiple times. Currently, our distributed autograd doesn't do that. Will fix that in a separate PR.

**Comment by mrshenli** [comment]
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28274876).

**Review by pritamdamania87 — APPROVED** [review]
Looks great, thanks for fixing this!

**Comment by mrshenli** [comment]
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28274876).

**Comment by mrshenli** [comment]
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28274876).

**Comment by mrshenli** [comment]
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28274876).

**Pull request closed without merging**